Qentelli test worklow

NidhiDixit09 opened the pull request:
## Reference

SDK-1874-Run tests from qentelli-saas-sdk-testing-automation repo
https://branch.atlassian.net/browse/SDK-1874

## Summary
Updated file .github/workflows/automation-trigger-test.yml to
* Build and upload static lib which will be used by Qentelli Automation Testbed App (for linking with framework build with the current branch).
* Build Qentelli Automation Testbed App 
* Creat and upload IPA 
* Check out code from qentelli-saas-sdk-testing-automation and copy uploaded IPA to testdata/builds/iOS_TestbedApp_21-Jun-2022-1.ipa
* Run tests and upload report
Note : Last job `buildAndRunAutomationTestSuite:`  is a copy-paste from https://github.com/BranchMetrics/qentelli-saas-sdk-testing-automation/blob/master/.github/workflows/ios-maven.yml. I have used same steps for running tests using maven.

## Motivation
Integrate Qentelli Automation Tests with iOS SDK.

## Type Of Change
- [ ] New feature (non-breaking change which adds functionality)

## Testing Instructions
Run workflow manually from the current Branch. Please see attached screen shot.
![image](https://user-images.githubusercontent.com/93544270/229188126-5ce1315a-c58b-48b7-b932-b781409cf5c9.png)


<!-- Checklist -->
<!-- My code follows the style guidelines of this project -->
<!-- I have performed a self-review of my code -->
<!-- I have commented my code, particularly in hard-to-understand areas -->
<!-- I have made corresponding changes to the documentation -->
<!-- I have added tests that prove my fix is effective or that my feature works -->
<!-- New and existing unit tests pass locally with my changes -->

cc @BranchMetrics/saas-sdk-devs for visibility.

echo-branch commented on .github/workflows/automation-trigger-test.yml:
```diff
@@ -1,18 +1,115 @@
+  # 3.Installs Certificates and Provisioning profile for signing App and creating IPA. Final IPA is uploaded as an artifact.
+  build-Automation-testbed-App:
+    runs-on: macos-latest
+    needs: [build-Brach-Static-framework]
```
Spelling error here. 
build-branch-static-framework

echo-branch commented on .github/workflows/automation-trigger-test.yml:
```diff
@@ -1,18 +1,115 @@
-  build:
-    runs-on: ubuntu-latest
+  # This job creates a static lib which will be used by Qentelli Automation Testbed App. Its uploaded as an artifact.
+  build-Brach-Static-framework:
```
spelling error here.
build-branch-static-framework

echo-branch commented on .github/workflows/automation-trigger-test.yml:
```diff
@@ -1,18 +1,115 @@
+          name: DeepLinkDemo.ipa
+          path: |
+                ./IPA/DeepLinkDemo.ipa
+  buildAndRunAutomationTestSuite:
```
let's rename the phases in a consistent manner. The first two use `-` names while this one uses camel case. I personally have no preference

echo-branch commented:
I'd like to rename the whole GHA from `.github/workflows/automation-trigger-test.yml` to something more informative and less verbose. Not sure what our options are here, for unit tests the GHA displays the PR title.

JagadeeshKaricherla-branch commented on .github/workflows/automation-trigger-test.yml:
```diff
@@ -1,18 +1,117 @@
+        run: |
+            ./scripts/prep_static_xcframework.sh
+      - name: Upload build artifacts
+        uses: actions/upload-artifact@v3
```
For the purpose of sharing data between jobs - unless there's a need to save/persist data after a workflow run, we can use cache via `actions/cache@v3` in place of upload and downloading artifacts.Using cache is in general faster than uploading and downloading artifacts.

NidhiDixit09 replied:
@JagadeeshKaricherla-branch As per my understanding, Cache will retrieve same files. Here files are changing with every run. Cache is used for re-using same files between different runs, so they are faster.  For example,  we are using cache here - https://github.com/BranchMetrics/ios-branch-deep-linking-attribution/blob/master/.github/workflows/release.yml for faster building of pods. Here it will not have any impact because they are re-generated with every run. Correct me if I am wrong.

JagadeeshKaricherla-branch commented on .github/workflows/automation-trigger-test.yml:
```diff
@@ -1,18 +1,117 @@
+          cp $PP_PATH ~/Library/MobileDevice/Provisioning\ Profiles
+          
+      - name: Download Branch SDK lib, build testbed app and create IPA File
+        uses: actions/download-artifact@v3
```
same as above. Use cache in place of artifacts.

JagadeeshKaricherla-branch commented on .github/workflows/automation-trigger-test.yml:
```diff
@@ -1,18 +1,117 @@
+            xcodebuild -workspace DeepLinkDemo.xcworkspace -scheme DeepLinkDemo  -sdk iphoneos archive -archivePath ./IPA/DeepLinkDemo.xcarchive
+            xcodebuild -exportArchive -archivePath ./IPA/DeepLinkDemo.xcarchive -exportOptionsPlist IPA/Info.plist -exportPath IPA/
+      - name: Upload build artifacts
+        uses: actions/upload-artifact@v3
```
Instead of uploading ipa as an artifact, we can do one of these : 
- use cache which is in general faster than uploading & downloading artifacts to share data between jobs
- upload the ipa to the browserstack cloud with a custom id or upload and save the returned `app_url`. This way we can break the browserstack test execution into independently executable / failable steps. We can fail fast in case of an issue.
https://www.browserstack.com/docs/app-automate/appium/upload-app-from-filesystem
https://www.browserstack.com/app-live/rest-api

JagadeeshKaricherla-branch commented on .github/workflows/automation-trigger-test.yml:
```diff
@@ -1,18 +1,117 @@
+        run: |
+           mvn clean test -DsuiteXMLFile=testng.xml -e -DInputFormat="json" -DInputFilePath="TestSelection_BrowserStack_iOS.json" -DAccessKey="${{ secrets.S3_ACCESS_KEY }}" -DSecretKey="${{ secrets.S3_SECRET_KEY }}" --log-file output.txt
+      - name: Print output
+        if: always()        # run this step even if previous step failed
```
use `success() || failure() `instead of `always()` as always causes the step to always execute, and returns true, even when `canceled`.
Also, always will execute even when any of the steps above in the job fail meaning not just the success or failure of `build_maven` step but steps above it and in that case, you won’t have output.txt so its better to check
`if: success() || steps. build_maven.outcome == 'failure'`

NidhiDixit09 commented:
@echo-branch Since your last approval, I did following updates in this PR -
 
* Updated workflow to point to master branch of SaaS SDK Automation Tests repo.
* Browserstack and AWS creds are added as repo secrets and they are passes to Tests as environment variables. So they are removed from command.
* Addressed review comments

gdeluna-branch commented on .github/workflows/automation-trigger-test.yml:
```diff
@@ -1,18 +1,122 @@
+      - name : Unzip and copy folder
+        run: |
+            ls -la ./
+            cp -R DeepLinkDemo.ipa testdata/builds/iOS_TestbedApp_21-Jun-2022-1.ipa
```
Let me know if you need help with uploading and setting the app_url env var
